chain/client/src/chunk_validation.rs

```diff
@@ -440,6 +401,101 @@ fn pre_validate_chunk_state_witness(
     })
 }
 
+/// Validate that receipt proofs contain the receipts that should be applied during the
+/// transition proven by ChunkStateWitness. The receipts are extracted from the proofs
+/// and arranged in the order in which they should be applied during the transition.
+/// TODO(resharding): Handle resharding properly. If the receipts were sent from before
+/// a resharding boundary, we should first validate the proof using the pre-resharding
+/// target_shard_id and then extract the receipts that are targeted at this half of a split shard.
+fn validate_source_receipt_proofs(
+    source_receipt_proofs: &HashMap<ChunkHash, ReceiptProof>,
+    blocks_after_last_last_chunk: &[Block],
+    last_chunk_block: &Block,
```

Comment:
Instead of having 2 separate arguments here I suggest to have single `receipt_source_blocks`. Then you can modify code in `pre_validate_chunk_state_witness` to directly construct it replacing `blocks_after_last_last_chunk` which is only used here.

Reply:
Great idea, I didn't realize that my function is the main consumer of the `after_last(_last)_chunk` vectors.
I changed the ranges covered by these vectors a bit and now the code is much cleaner :)